src/main/java/com/shapesecurity/salvation/data/Policy.java

```diff
@@ -195,20 +190,70 @@ private void optimise() {
             }
 
             // * remove unnecessary default-src directives if all source directives exist
-            if (this.directives.containsKey(ScriptSrcDirective.class) &&
-                this.directives.containsKey(StyleSrcDirective.class) &&
-                this.directives.containsKey(ImgSrcDirective.class) &&
-                this.directives.containsKey(ChildSrcDirective.class) &&
-                this.directives.containsKey(ConnectSrcDirective.class) &&
-                this.directives.containsKey(FontSrcDirective.class) &&
-                this.directives.containsKey(MediaSrcDirective.class) &&
-                this.directives.containsKey(ObjectSrcDirective.class) &&
-                this.directives.containsKey(ManifestSrcDirective.class)) {
+            if (fold(Directive.getFetchDirectives(), (memo, fetchDirective) -> memo && this.directives.containsKey(fetchDirective), true)) {
                 this.directives.remove(DefaultSrcDirective.class);
             }
         }
     }
 
+    private <A, B> B fold(List<A> list, BiFunction<B, A, B> reducer, B init) {
+        for (A a : list) {
+            init = reducer.apply(init, a);
+        }
+        return init;
+    }
+
+    public void postProcessOptimisation() {
+
+        DefaultSrcDirective defaultSrcDirective;
+
+        int directiveCount = this.directives.size();
+
+        // below optimisations kick in when there are two or more directives
+        if (directiveCount < 2) {
+            return;
+        }
+        // move source-expressions to default-src if they exist in every fetch directive
+        SourceListDirective prevDirective = null;
+        int fetchDirectiveCount = 0;
+        boolean isfetchSourceListIdentical = true;
+        for (Map.Entry<Class<?>, Directive<? extends DirectiveValue>> entry : this.directives.entrySet()) {
+            Directive<? extends DirectiveValue> directive = entry.getValue();
+            if (directive instanceof FetchDirective) {
+                fetchDirectiveCount++;
+                if (prevDirective != null) {
+                    if (!prevDirective.sourceListEquals(directive)) {
+                        isfetchSourceListIdentical = false;
+                        break;
+                    }
+                }
+                prevDirective = (SourceListDirective) directive;
+            }
+        }
+
+        // remove all fetch directives and replace with default-src
+        if (fetchDirectiveCount == Directive.getFetchDirectives().size() && isfetchSourceListIdentical) {
```

Comment:
`Directive.getFetchDirectives().size()` can be made into a constant on `Directive` statically.